### PR TITLE
Type checking

### PR DESCRIPTION
**Pull request by FilipKon13**
Initial version to allow an early feedback - no diagnostics atm, the tests do not cover checking errors

### CODE REVIEW

File: src/test/kotlin/cacophony/semantic/TypeCheckerTest.kt [hunk 1]
```diff
@@ -0,0 +1,952 @@
...
+    private val intLiteral = Literal.IntLiteral(lc, 7)
+    private val booleanLiteral = Literal.BoolLiteral(lc, true)
+
+    @Test
+    fun `ok - empty block`() {
+        val ast = Block(lc, emptyList())
+        val diagnostics = getDiagnostic()
+        val result = checkTypes(ast, diagnostics, emptyMap())
+        assertTypeEquals(BuiltinType.UnitType, result[ast])
+        assertNull(diagnostics.msg)
```
Comment (km1chno) [inline src/test/kotlin/cacophony/semantic/TypeCheckerTest.kt:72]: you can do

```
@MockK
lateinit var diagnostics: Diagnostics

@BeforeEach
fun setUpMocks() {
    MockKAnnotations.init(this, relaxUnitFun = true)
    every { diagnostics.report(any(), any<Location>()) } just runs
}
```

and then in tests

```
verify(exactly = 0) { diagnostics.report(any(), any<Location>()) }
```

like in https://github.com/mhorod/cacophony/blob/53493d26df99c0c86e1fb7beedbe9c2536f554aa/src/test/kotlin/cacophony/lexer/RegularLanguageLexerTest.kt#L98

Comment (FilipKon13) [reply src/test/kotlin/cacophony/semantic/TypeCheckerTest.kt:70]: I can, but should I?

Comment (km1chno) [reply src/test/kotlin/cacophony/semantic/TypeCheckerTest.kt:70]: up to you, I prefer it this way but its a matter of taste

File: src/main/kotlin/cacophony/semantic/TypeChecker.kt [hunk 2]
```diff
@@ -1,19 +1,345 @@
...
+                is OperatorUnary.Negation -> {
+                    val innerType = typeExpression(expression.expression) ?: return null
+                    if (!isSubtype(innerType, BuiltinType.BooleanType)) {
+                        diagnostics.report("Type $innerType does not support unary ! operator", expression.expression.range)
+                        return null
+                    }
+                    BuiltinType.BooleanType
+                }
+                // TODO: who should check if break is inside loop?
+                is Statement.BreakStatement -> BuiltinType.UnitType
```
Comment (MisieqqeisiM) [inline src/main/kotlin/cacophony/semantic/TypeChecker.kt:201]: break should have Void type

**Comment by FilipKon13** [comment]
There are some TODOs refering to code duplication - I intend to remove those before the merge. Additionally, there is inconsistency in how name of function type is generated, but it does not show up atm because we do not have function as args - I will fix this as well

File: src/main/kotlin/cacophony/semantic/TypeChecker.kt (hunk 2 above)
Comment (FilipKon13) [reply src/main/kotlin/cacophony/semantic/TypeChecker.kt:201]: I will add test for this as well

**Review by MisieqqeisiM — APPROVED** [review]
I like it

**Pull request merged**